### ISSUE TITLE
Unable to correct drift with `pulumi up --refresh` after external changes

### ISSUE THREAD

**Issue by rquitales** [pulumi/pulumi-kubernetes#2404]
### What happened?

I'm unable to correct drift in Kubernetes resources when running `pulumi up --refresh`, or `pulumi refresh; pulumi up` after the resources on the cluster have been modified with another tool/process. I have tested this with both CSA and SSA, and both seem to have issues with how they are handled.

In CSA, the resources on the cluster are never updated, but `pulumi refresh` indicates a diff that doesn't cause `pulumi up` to trigger a diff to run.

In SSA, with a ConfigMap, the resource is recreated which may not be ideal if we're using auto-generated names and the name of the CM will also be recreated. With changes to a deployment spec however, `pulumi up` updates in place. The handling of both CM and deployments should be consistent in that they are both updated in place.

### Expected Behavior

`pulumi up --refresh` should be able to correct any spec changes to Kubernetes resources made externally when using CSA. When using SSA, it should not recreate the ConfigMap object.

### Steps to reproduce

I've added 2 test cases that show cases this behaviour, one each for CSA and SSA in #2403.
Test run: https://github.com/pulumi/pulumi-kubernetes/actions/runs/4966037016/jobs/8887293231?pr=2403
```
FAIL: TestClientSideDriftCorrectCSA
FAIL: TestClientSideDriftCorrectSSA
```

#### Manual Reproduction Steps:

1. `pulumi new kubernetes-typescript`
2. Add a new ConfigMap resource with some data eg: `foo: bar`
3. `pulumi up --yes`
4. Edit the ConfigMap and/or deployment (eg `kubectl patch cm <cm-name> -p '{"data": {"foo": "newValue"}}'`)
5. `pulumi up --refresh --yes`
6. Get the ConfigMap or deployment object and note that they are not updated for CSA (`kubectl get cm <cm-name>`)

### Output of `pulumi about`

CLI
Version      3.67.0
Go Version   go1.20.4
Go Compiler  gc

Plugins
NAME        VERSION
kubernetes  3.27.1
nodejs      unknown

Host
OS       darwin
Version  13.3.1
Arch     arm64

This project is written in nodejs: executable='/opt/homebrew/bin/node' version='v19.6.0'

TYPE                           URN
pulumi:pulumi:Stack            urn:pulumi:dev::k8s-resourceVersions::pulumi:pulumi:Stack::k8s-resourceVersions-dev
pulumi:providers:kubernetes    urn:pulumi:dev::k8s-resourceVersions::pulumi:providers:kubernetes::k8s
kubernetes:apps/v1:Deployment  urn:pulumi:dev::k8s-resourceVersions::kubernetes:apps/v1:Deployment::nginx

Found no pending operations associated with dev

Dependencies:
NAME                 VERSION
@pulumi/kubernetes   3.27.1
@pulumi/kubernetesx  0.1.6
@pulumi/pulumi       3.67.0
@types/node          16.18.29

### Additional context

I have not looked into the root cause yet.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by lblackstone** [comment]
Could this be the same issue as https://github.com/pulumi/pulumi-kubernetes/issues/694? I can't recall if `kubectl patch` updates the `lastAppliedConfiguration` annotation, but that affects CSA diffing. 

I'm reviewing the linked PR now, and it looks like the code changes are to relatively old parts of the code. The most recent edit I saw was from August 2022, so I'd be surprised if this was a regression. It may be incorrect, but if so, I think it's been that way for a long time.

**Comment by rquitales** [comment]
Yes, this behaviour has been in the provider for a long time, and the linked issue is of the same behaviour. `lastAppliedConfiguration` is only updated by `kubectl apply`, so this issue would persist for `kubectl patch`, or other controllers/tooling that modify the resources.

For now, I'll downgrade this issue from a P1 since it seems like a long standing issue we've had, and it doesn't appear to be a pressing concern reported by other users.

The next steps would be to expand our test suites so this behaviour change can be merged with more confidence.
